### PR TITLE
Implemented mechanism of dynamic change of app settings

### PR DESCRIPTION
**Pull request by vsakharevych**

### CODE REVIEW

File: src/documentmanager/documentmanager.cpp [hunk 1]
```diff
@@ -298,7 +308,7 @@ QMdiSubWindow* DocumentManager::openedDoc(const QString &fileName)
         openedDocIter = std::find_if(subWdwList.cbegin(), subWdwList.cend(),
                                      [&fileName](const auto& doc)
         {
-            return static_cast<CodeEditor*>(doc->widget())->getFileName() == fileName;
+            return qobject_cast<CodeEditor*>(doc->widget())->getFileName() == fileName;
```
Comment (PeterLysyk) [inline src/documentmanager/documentmanager.cpp]: qobject_cast can return nullptr. after that you're trying to call method getFileName. If you call this fucntion to nullptr it'll crush the program. Check what cast returned before calling the method

File: src/documentmanager/documentmanager.cpp [hunk 2]
```diff
@@ -325,43 +333,39 @@ QMdiArea* DocumentManager::lastAreaInFocus()
     // search for area which accomodates last doc in focus
     for (const auto& area: mDocAreas)
```
Comment (PeterLysyk) [inline src/documentmanager/documentmanager.cpp]: `for (const auto& area: mDocAreas)`
rewrite it to
`for (const auto &area: mDocAreas)`

File: src/documentmanager/documentmanager.cpp [hunk 3]
```diff
@@ -325,43 +333,39 @@ QMdiArea* DocumentManager::lastAreaInFocus()
...
     {
-        auto subWdwList = area->subWindowList();
-
-        areaInFocusIter = std::find_if(subWdwList.cbegin(), subWdwList.cend(),
-                                       [this](const auto& doc)
-        {
-            return static_cast<CodeEditor*>(doc->widget()) == mpPrevEditorInFocus;
-        });
+        auto currentWdw = area->currentSubWindow();
```
Comment (PeterLysyk) [inline src/documentmanager/documentmanager.cpp]: auto **currentWdw** = area->currentSubWindow();
I suppose **currentWindow** is better name for variable

File: src/documentmanager/documentmanager.cpp [hunk 4]
```diff
@@ -325,43 +333,39 @@ QMdiArea* DocumentManager::lastAreaInFocus()
...
 
-        if (areaInFocusIter != subWdwList.end())
+        if (currentWdw)
         {
-            return area;
+            auto doc = qobject_cast<CodeEditor*>(currentWdw->widget());
+            if (doc && doc == mpPrevEditorInFocus)
```
Comment (PeterLysyk) [inline src/documentmanager/documentmanager.cpp]: `if (doc && doc == mpPrevEditorInFocus)`
in accordance to priority of operations, rewrite it to
`if (doc && (doc == mpPrevEditorInFocus))`

File: src/documentmanager/documentmanager.cpp [hunk 5]
```diff
@@ -381,7 +385,7 @@ QMdiArea* DocumentManager::getArea(CodeEditor *doc)
         areaIter = std::find_if(subWdwList.cbegin(), subWdwList.cend(),
                                 [&doc](const auto& document)
         {
-            return static_cast<CodeEditor*>(document->widget()) == doc;
+            return qobject_cast<CodeEditor*>(document->widget()) == doc;
```
Comment (PeterLysyk) [inline src/documentmanager/documentmanager.cpp]: `qobject_cast<CodeEditor*>(document->widget()` can return nullptr and after that you're comparing its widget with `doc`. Check it

File: src/documentmanager/documentmanager.cpp [hunk 6]
```diff
@@ -182,6 +184,14 @@ void DocumentManager::loadFile(CodeEditor *newView, const QString &fileName)
 
 void DocumentManager::onSplit(Qt::Orientation orientation)
 {
+    // if there is only one doc area -
```
Comment (PeterLysyk) [inline src/documentmanager/documentmanager.cpp]: it can be 0 doc area

File: src/documentmanager/documentmanager.cpp [hunk 7]
```diff
@@ -325,43 +333,39 @@ QMdiArea* DocumentManager::lastAreaInFocus()
...
     return nullptr;
 }
 
-QMdiArea *DocumentManager::areaInFocus()
+QMdiArea* DocumentManager::areaInFocus()
 {
-    QList<QMdiSubWindow*>::const_iterator areaInFocusIter;
-
     // search for area which accomodates current doc in focus
     for (const auto& area: mDocAreas)
```
Comment (PeterLysyk) [inline src/documentmanager/documentmanager.cpp]: for (const auto **&**area: mDocAreas)

Comment (PeterLysyk) [inline src/documentmanager/documentmanager.cpp]: `for (const auto &area: mDocAreas)`

File: src/documentmanager/documentmanager.cpp [hunk 8]
```diff
@@ -414,7 +418,7 @@ CodeEditor* DocumentManager::getCurrentDocument()
     // receive current sub wdw from area in focus
     // if current sub wdw is null - we return null to indicate search failure
     auto pCurrentDocument = pAreaInFocus->currentSubWindow();
-    return pCurrentDocument ? static_cast<CodeEditor*>(pCurrentDocument->widget()) : nullptr;
+    return pCurrentDocument ? qobject_cast<CodeEditor*>(pCurrentDocument->widget()) : nullptr;
```
Comment (PeterLysyk) [inline src/documentmanager/documentmanager.cpp:423]: check nullptr again

File: src/documentmanager/documentmanager.cpp [hunk 9]
```diff
@@ -532,27 +545,42 @@ void DocumentManager::closeEmptyDocArea()
     }
 }
 
-void DocumentManager::closeCurrentDocArea()
+void DocumentManager::configureDocuments(std::function<void(DocumentManager*, CodeEditor*, const QString&)> functor,
+                                        const QString &newValue)
 {
-    if (mDocAreas.size() < 2)
+    // applying new settings to every opened doc view
+    for (const auto& area: mDocAreas)
```
Comment (PeterLysyk) [inline src/documentmanager/documentmanager.cpp]: for (const auto &area: mDocAreas)

File: src/documentmanager/documentmanager.cpp [hunk 10]
```diff
@@ -532,27 +545,42 @@ void DocumentManager::closeEmptyDocArea()
...
-
-    // close area in focus
-    auto pAreaInFocus = areaInFocus();
+        auto windowsList = area->subWindowList();
 
-    if (!pAreaInFocus)
-    {
-        return;
+        std::for_each(windowsList.begin(), windowsList.end(),
+                      [this, &functor, &newValue](const auto& wdw)
```
Comment (PeterLysyk) [inline src/documentmanager/documentmanager.cpp]: [this, &functor, &newValue](const auto &wdw)

File: src/editor/autocodecompleter.cpp [hunk 11]
```diff
@@ -0,0 +1,70 @@
+#include "autocodecompleter.h"
```
Comment (PeterLysyk) [inline src/editor/autocodecompleter.cpp:1]: This file is old version which appered here accidentally. Don't read it)

File: src/mainwindow/mainwindow.cpp [hunk 12]
```diff
@@ -561,14 +553,15 @@ void MainWindow::onConnectTriggered()
...
     }
-    mCurrentUserName = userInput;
     mpChatWindowDock->setUserName(userInput);
-    mplocalConnector->configureOnLogin(mCurrentUserName);
+    mplocalConnector->configureOnLogin(userInput);
 }
 
 void MainWindow::onSettingsTriggered()
 {
     MenuOptions * menuOptions = new MenuOptions(this);
```
Comment (PeterLysyk) [inline src/mainwindow/mainwindow.cpp:562]: MenuOptions *menuOptions = new MenuOptions(this);

File: src/mainwindow/mainwindow.cpp [hunk 13]
```diff
@@ -624,6 +617,17 @@ void MainWindow::onConnectionStatusChanged(bool status)
     }
 }
 
+void MainWindow::onSettingsChanged(std::map<QString, QString> newValues)
+{
+    SettingsConfigurator settingsConfigurator;
+
+    for (const auto& newValue: newValues)
```
Comment (PeterLysyk) [inline src/mainwindow/mainwindow.cpp:624]: for (const auto &newValue: newValues)

File: src/mainwindow/mainwindow.cpp [hunk 14]
```diff
@@ -696,13 +710,45 @@ void MainWindow::restoreMainWindowState()
...
 {
     // fusion style is applied globally
     // if platform does not support fusion, default style is applied
     qApp->setStyle(QStyleFactory::create("Fusion"));
+
+    // restores style palette set in previous app session
+    QSettings savedSettings(QApplication::organizationName(), QApplication::applicationName());
+    QString styleName = {savedSettings.contains("style") ?
+                         savedSettings.value("style").toString()
+                         : "WHITE"};
```
Comment (PeterLysyk) [inline src/mainwindow/mainwindow.cpp:724]: it's not obvious that 
`savedSettings.contains("style") ?
                         savedSettings.value("style").toString()
                         : "WHITE"}`
are default settings. comment stm like this

File: src/mainwindow/mainwindow.cpp [hunk 15]
```diff
@@ -696,13 +710,45 @@ void MainWindow::restoreMainWindowState()
...
+    setAppStyle(styleName);
+}
+
+void MainWindow::setAppStyle(const QString &styleName)
+{
     // dark style palette is created & set globally
-    QPalette palette = mpPaletteConfigurator->getPalette("DARK");
+    QPalette palette = mpPaletteConfigurator->getPalette(styleName);
     //QPalette newPal = palette();
```
Comment (PeterLysyk) [inline src/mainwindow/mainwindow.cpp]: delete commented code

File: src/settings/menuoptions.h [hunk 16]
```diff
@@ -54,6 +54,7 @@ class MenuOptions : public QDialog
     int mMenuHeight             = 280;
 private slots:
     void onBtnClicked(QAbstractButton * button);
```
Comment (PeterLysyk) [inline src/settings/menuoptions.h:56]: void onBtnClicked(QAbstractButton *button);

File: src/settings/menuoptions.h [hunk 17]
```diff
@@ -54,6 +54,7 @@ class MenuOptions : public QDialog
     int mMenuHeight             = 280;
```
Comment (PeterLysyk) [inline src/settings/menuoptions.h:54]: why this space is so large?

File: src/documentmanager/documentmanager.cpp (hunk 1 above)
Comment (vsakharevych) [reply src/documentmanager/documentmanager.cpp]: fixed

File: src/documentmanager/documentmanager.cpp (hunk 2 above)
Comment (vsakharevych) [reply src/documentmanager/documentmanager.cpp]: fixed

File: src/documentmanager/documentmanager.cpp (hunk 3 above)
Comment (vsakharevych) [reply src/documentmanager/documentmanager.cpp]: fixed

File: src/documentmanager/documentmanager.cpp (hunk 4 above)
Comment (vsakharevych) [reply src/documentmanager/documentmanager.cpp]: fixed

File: src/documentmanager/documentmanager.cpp (hunk 5 above)
Comment (vsakharevych) [reply src/documentmanager/documentmanager.cpp]: fixed

File: src/documentmanager/documentmanager.cpp (hunk 6 above)
Comment (vsakharevych) [reply src/documentmanager/documentmanager.cpp]: provided additional comment to clarify current block of code

File: src/documentmanager/documentmanager.cpp (hunk 7 above)
Comment (vsakharevych) [reply src/documentmanager/documentmanager.cpp]: fixed

File: src/documentmanager/documentmanager.cpp (hunk 8 above)
Comment (vsakharevych) [reply src/documentmanager/documentmanager.cpp:423]: if qobject_cast fails & returns null_ptr then null_ptr is returned from method. Return value is checked by calling method.
Current method was designed that way

File: src/documentmanager/documentmanager.cpp (hunk 9 above)
Comment (vsakharevych) [reply src/documentmanager/documentmanager.cpp]: fixed

File: src/documentmanager/documentmanager.cpp (hunk 10 above)
Comment (vsakharevych) [reply src/documentmanager/documentmanager.cpp]: fixed

File: src/mainwindow/mainwindow.cpp (hunk 12 above)
Comment (vsakharevych) [reply src/mainwindow/mainwindow.cpp:562]: It is Roman's code

File: src/mainwindow/mainwindow.cpp (hunk 14 above)
Comment (vsakharevych) [reply src/mainwindow/mainwindow.cpp:724]: added comments

File: src/mainwindow/mainwindow.cpp (hunk 15 above)
Comment (vsakharevych) [reply src/mainwindow/mainwindow.cpp]: fixed

File: src/settings/menuoptions.h (hunk 16 above)
Comment (vsakharevych) [reply src/settings/menuoptions.h:56]: it is Roman's code

File: src/settings/menuoptions.h (hunk 17 above)
Comment (vsakharevych) [reply src/settings/menuoptions.h:54]: it is Roman's code

File: src/documentmanager/documentmanager.cpp (hunk 8 above)
Comment (PeterLysyk) [reply src/documentmanager/documentmanager.cpp:423]: see it, ok.

File: src/editor/autocodecompleter.cpp [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+#include<QDebug>
+
+AutoCodeCompleter::AutoCodeCompleter(const QStringList &completions, QObject *parent):
+    QCompleter(completions, parent)
+{
+    connect(this, SIGNAL(activated(QString)), this, SLOT(replaceCurrentWord(QString)));
```
Comment (MarsLviv) [inline src/editor/autocodecompleter.cpp:7]: use new style for connection or lamda

File: src/editor/codeeditor.h [hunk 19]
```diff
@@ -23,6 +23,9 @@ const int TOP_UNUSED_PIXELS_HEIGHT = 4;
 #include<QVector>
 #include<QFont>
 #include<QStatusBar>
+#include<QStringList>
+#include<QCompleter>
+#include"autocodecompleter.h"
 
 
 class QPaintEvent;
```
Comment (MarsLviv) [inline src/editor/codeeditor.h:31]: use QT_BEGIN_NAMESPACE and QT_END_NAMESPACE to enclose class declaration

File: src/mainwindow/mainwindow.cpp [hunk 20]
```diff
@@ -12,6 +12,7 @@
 #include <QFile>
 
 #include "localconnectorgenerator.h"
+#include "settingsconfigurator.h"
 #include "paletteconfigurator.h"
 #include "projectviewerdock.h"
 #include "documentmanager.h"
```
Comment (MarsLviv) [inline src/mainwindow/mainwindow.cpp:18]: use QT_END_NAMESPACE and QT_BEGIN_NAMESPACE

**Review by PeterLysyk — APPROVED** [review]

**Review by MarsLviv — APPROVED** [review]

File: src/editor/autocodecompleter.cpp (hunk 18 above)
Comment (vsakharevych) [reply src/editor/autocodecompleter.cpp:7]: It is Petro's code

File: src/editor/codeeditor.h (hunk 19 above)
Comment (vsakharevych) [reply src/editor/codeeditor.h:31]: It is Petro's code

File: src/mainwindow/mainwindow.cpp (hunk 20 above)
Comment (vsakharevych) [reply src/mainwindow/mainwindow.cpp:18]: It is cpp file

**Pull request merged**